### DISCUSSION TITLE
Feature parity with Amethyst

### DISCUSSION THREAD

**Discussion by inodentry** [bevyengine/bevy#1345] [Ideas]
Bevy is already "ahead" of Amethyst in some ways, such as, notably, the ECS ergonomics that made Bevy popular. Bevy is also lacking many useful game engine features.

While I don't want this to be treated as a competition, I want to see Bevy offer useful functionality for users in a wide variety of areas, and I think we should strive to improve on what other game engines offer.

Therefore, I think keeping a list of the areas in which Bevy is currently lacking in comparison to Amethyst would be helpful for Bevy development.

Let's use this discussion to track the features Bevy would need "to catch up" to Amethyst, functionality-wise, and file github issues or discussion posts to track each one specifically.

Here are some that I could find:

 - Debug rendering (wireframe, arbitrary lines in 3d space): #409.
 - Localization support (Amethyst has some primitive integration with `Fluent`): #461. We should offer more than just text/ui localization; see the linked issue. [Amethyst](https://docs.amethyst.rs/stable/amethyst_locale/index.html)
 - Easily load app-specific arbitrary `serde`-deserializable data (like configs, etc.) from asset files, without having to write boilerplate for a custom asset loader: #1265. Perhaps the ongoing Distill (formerly `atelier-assets`, from Amethyst) integration (#708) will help with this?
 - Animation (skeletal and otherwise): #91 #95. [Amethyst](https://docs.amethyst.rs/stable/amethyst_animation/index.html)
 - Networking: Amethyst has a general abstraction to be backed by protocol backends. They also developed [laminar](https://github.com/amethyst/laminar). For Bevy, there are external crates to experimentally integrate [laminar](https://github.com/ncallaway/bevy_prototype_networking_laminar), [turbulence](https://github.com/smokku/bevy_networking_turbulence), [simple tcp/udp](https://github.com/0x22fe/bevy_prototype_simple_net).
 - Setting the application/window icon (displayed by the OS in the taskbar, etc.): #1163 #1031. [Amethyst](https://docs.amethyst.rs/stable/amethyst_window/struct.Icon.html)
 - Mapping of input controls to game actions (remapping, keybindings): [`Kurinji`](https://github.com/PradeepKumarRajamanickam/kurinji) is an unofficial bevy crate with some preliminary work on this.
 - Frame rate limiting / throttling refreshes: #1343.
 - Audio system improvements (volume controls, mixing, many sound sources / with world entities acting as sound emitters): Perhaps the ongoing Kira integration (#1304) is the way to go here.
 - Support for using fonts installed in the operating system / default font: #1325 #1017 ; we might want to integrate [`font-kit`](https://github.com/servo/font-kit).
 - UI text input support. This will be required as part of the Editor-Ready UI focus area: #254.
 - UI nine-patch support. Some preliminary work available in the unofficial [`bevy_ninepatch`](https://github.com/mockersf/bevy_extra/tree/master/bevy_ninepatch) crate.
 - Physically-Based Rendering. This is one of our focus areas: #179. Existing work: #1160 #261 .
 - Camera controls (fly camera, arcball, etc.). Currently available as [unofficial crates](https://github.com/bevyengine/awesome-bevy#3d), there is a (semi-broken) example in the [cheatbook](https://bevy-cheatbook.github.io/cookbook/pan-orbit-camera.html). These should probably be kept as external crates.
 - In-game / onscreen FPS counter (and other performance diagnostics). A simple FPS counter is available as an [example](https://github.com/bevyengine/bevy/blob/master/examples/ui/text.rs).
 - Asset loading progress tracking. Useful for "loading screens" / "progress bars".
 
 ... more?

**Comment by alice-i-cecile** [comment]
* Networking is seriously lacking, although I'm not sure other Rust game engines are any further along.
* A pipeline for deployment to end users on various platforms is also critical for pushing this into production.
* It's currently possible to create particle effects, but it's neither ergonomic or performant.
* States (as of 0.4) are frustrating and confusing to use.
* Tools to control frame rate are confusing and imperfect.
* Saving and loading games is not well explored, and Scene technology is not fully adequate.
* There's no clear, ergonomic way to pause and resume games without risking lost data due to change detection and events.

**Reply by inodentry** [reply]
> particle effects

Does amethyst have a particle system? I can't find anything.

(This discussion is specifically for tracking where we stand in comparison to Amethyst, not a general wishlist.)

> A pipeline for deployment to end users on various platforms is also critical for pushing this into production.

Do we have any existing issues or work done towards this? What does Amethyst offer?

> Networking

edited into the OP, thanks!

**Reply by cart** [reply]
> There's no clear, ergonomic way to pause and resume games without risking lost data due to change detection and events.

I think pausing should be a "state", which is already possible with StateStage.

**Reply by alice-i-cecile** [reply]
> I think pausing should be a "state", which is already possible with StateStage.

I agree with the former, but not the latter :) I've included my explanation in #1353, rather than derailing in here.

**Comment by kragil** [comment]
- a canvas

**Comment by cart** [comment]
I'm not specifically interested in driving a "Amethyst feature parity" list down to zero. Its worth having a discussion about "useful Amethyst features", creating new issues for those features, and prioritizing them accordingly. Thats also worth doing for other engines.

But making Bevy a "drop in replacement" for Amethyst specifically is a non-goal.

**Reply by inodentry** [reply]
Absolutely agreed. This was exactly the intention of this post. I want to identify areas in which we are lacking, using Amethyst as inspiration in this case. The intention was not to suggest to exactly match/imitate/replace Amethyst. I think we should learn from them and try to do better (and same goes for other game engines).

> Thats also worth doing for other engines.

Totally. I am personally not really familiar with any of them, though, so someone more knowledgeable would have to look into that. Though, I think we should delay comparisons with the "big guys" (unreal/unity/etc.) until we are at least at the stage where we are making a Bevy Editor.

---

I also think there is value in us caring about comparisons with Amethyst specifically, because Amethyst is the other "big Rust game engine project", so naturally, people (outsiders) want to compare the two.
